### PR TITLE
Quickly Launch Different Chrome for Privacy Sandbox Demo

### PR DESCRIPTION
**Pull request by MiteshShah**
## Description

This script helps to launch different Chrome ( Macintosh and Linux platforms) quickly for Privacy Sandbox Demo.

## Testing Instructions

```
curl -sL https://rt.cx/psat | bash   
```

Above command call `install.sh` which download `chrome_launcher.sh` and source it into `.zshrc` or `.bashrc` also automatically source into the current terminal session as well.

### CODE REVIEW

File: chrome_launcher.sh [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    }
+fi
+
+# Loads Chrome with a temporary profile that is deleted after Chrome is closed
+chrome-default() {
+  launch_chrome \
+    --install-autogenerated-theme='150,220,150'
+}
+
+chrome-3pcd() {
```
Comment (dhsathiya) [inline chrome_launcher.sh:47]: @MiteshShah can you please add one liner before each chrome function for it's purpose? e.g. line number [41](https://github.com/GoogleChromeLabs/ps-analysis-tool/pull/211/files#diff-89ef72b5c5b4c0decec968e6b38afaffaaabecc1db74269094fd8b0f65f84c51R41)

**Comment by sayedtaqui** [comment]
Instead of downloading the script, shouldn't they be able to run using npm script ? Something like this?

In `package.json`
```json
"scripts": {
	"launch-chrome:default": "bash ./bin/chrome_launcher/chrome-default.sh",
	"launch-chrome:default-ps": "bash ./bin/chrome_launcher/chrome-default-ps.sh",
	"launch-chrome:3pcd": "bash ./bin/chrome_launcher/chrome-3pcd.sh",
	"launch-chrome:3pcd-ps": "bash ./bin/chrome_launcher/chrome-3pcd-ps.sh",
	"launch-chrome:chip": "bash ./bin/chrome_launcher/chrome-chip.sh"		
}
```

and then run 

```bash
npm run launch-chrome:default
npm run launch-chrome:default-ps
npm run launch-chrome:3pcd
...
```

**Comment by gagan0123** [comment]
@sayedtaqui 

If we go with `npm` based commands, it will require the users to install git, clone the code, install node, install npm, then install npm dependencies, then they'll be able to run the commands. While going with directly having the script, they only need to run the script once, whether they have git, node or npm installed, it won't matter.

**Comment by sayedtaqui** [comment]
> @sayedtaqui
> 
> If we go with `npm` based commands, it will require the users to install git, clone the code, install node, install npm, then install npm dependencies, then they'll be able to run the commands. While going with directly having the script, they only need to run the script once, whether they have git, node or npm installed, it won't matter.

Well, we do expect the user to have those for running CLI commands (example npm run cli -- -u <https://bbc.com>). However, if we are only expecting them to download the file, it doesn't make sense to keep it in the repo because it has no relation to the rest of the codebase. Instead, it can simply be a link to a gist file in the readme. Although, I still think it would be helpful to include the npm command as well.

**Comment by amedina** [comment]
Let's go with the simple command to start and we may add the npm script as
well later; although the latter may not be needed.

**Comment by gagan0123** [comment]
@MiteshShah 

`/usr/local/bin` is owned by the user `root` so the script won't be able to write data directly. For writing there we'll need to add `sudo` when downloading the file.

**Comment by gagan0123** [comment]
Also, can we have a wrapper around to check whether the file exists or not before sourcing it in the `.zshrc` and `.bashrc`, something like this:
```
# Chrome profiles
if [ -f ~/Tools/Scripts/chrome.sh ]; then
    source ~/Tools/Scripts/chrome.sh
fi
```

**Comment by MiteshShah** [comment]
@gagan0123  I'd modify script so its wont need `sudo` anymore

To test please clear old source line from `.bashrc` `.zshrc` and run following command again

```
curl -sL https://rt.cx/gpsat | bash   
```

Its will create `~/bin` directory and add our chrome_launcher.sh inside it so wont needed `sudo`
Also now if you check your `.bashrc` or `.zshrc` its also check if file exist before source it like you mentioned on above comments.

**Comment by sayedtaqui** [comment]
@MiteshShah Can you please move these files in a new `bin` folder.

**Comment by gagan0123** [comment]
@MiteshShah

Thanks, now its loading the correct install script, the redirection was cached in my browser so I was still seeing the old install script.

 
 
@sayedtaqui 

Can you provide more information where to create a new `bin` directory and why?

Do you mean in the repo, we should create a `bin` directory and place these scripts there?

File: chrome_launcher.sh [hunk 2]
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
+
+# Download Extension
+extension_setup() {
+  ps_analysis_tool_version=v0.3.0
```
Comment (gagan0123) [inline chrome_launcher.sh:5]: @MiteshShah 0.3.1 has been released, can we please change this to 0.3.1?

**Review by gagan0123 — CHANGES_REQUESTED** [review]

**Comment by sayedtaqui** [comment]
@gagan0123 

> Can you provide more information where to create a new bin directory and why?

> Do you mean in the repo, we should create a bin directory and place these scripts there?

Yes, please create a `bin` folder in the root of the repository and move these `.sh` files into that folder. These `.sh` files should not be directly dumped in the root folder.

File: chrome_launcher.sh (hunk 2 above)
Comment (MiteshShah) [reply chrome_launcher.sh:5]: on it!

**Comment by MiteshShah** [comment]
@sayedtaqui Moved script to `bin` folder
@gagan0123 v0.3.1 is updated as well.

File: chrome_launcher.sh (hunk 2 above)
Comment (MiteshShah) [reply chrome_launcher.sh]: done

**Review by gagan0123 — APPROVED** [review]

File: bin/chrome_launcher.sh [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    --test-third-party-cookie-phaseout \
+    --enable-features="FirstPartySets,StorageAccessAPI,StorageAccessAPIForOriginExtension,PageInfoCookiesSubpage,PrivacySandboxFirstPartySetsUI"
+}
+
+# Launch default chrome with Privacy Sandbox extension
+chrome-default-ps() {
+  extension_setup
+  launch_chrome \
+    --install-autogenerated-theme='150,220,150' \
+    --load-extension="/tmp/ps-analysis-tool-v0.3.0/extension/"
```
Comment (gagan0123) [inline bin/chrome_launcher.sh:60]: Modify it to get the version from the `ps_analysis_tool_version` variable that we defined

File: bin/chrome_launcher.sh [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+# Launch default chrome with 3rd Party Cookie Phaseout with Privacy Sandbox extension
+chrome-3pcd-ps() {
+  extension_setup
+  launch_chrome \
+    --install-autogenerated-theme='255,51,51' \
+    --test-third-party-cookie-phaseout \
+    --load-extension="/tmp/ps-analysis-tool-v0.3.0/extension" \
```
Comment (gagan0123) [inline bin/chrome_launcher.sh:69]: Modify it to get the version from the `ps_analysis_tool_version` variable that we defined

**Review by gagan0123 — CHANGES_REQUESTED** [review]

File: bin/chrome_launcher.sh (hunk 4 above)
Comment (MiteshShah) [reply bin/chrome_launcher.sh]: Done

File: bin/chrome_launcher.sh (hunk 3 above)
Comment (MiteshShah) [reply bin/chrome_launcher.sh]: Done

File: bin/chrome_launcher.sh [hunk 5]
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
+
+# Download Extension
+extension_setup() {
+  ps_analysis_tool_version=v0.3.1
+  cd /tmp/
```
Comment (gagan0123) [inline bin/chrome_launcher.sh:6]: When we do `cd /tmp/` the directory is changed to that, and does not revert back to where we started, can we please ensure that after downloading and setting up the extension, we go back to the directory where we were?

**Review by gagan0123 — CHANGES_REQUESTED** [review]

**Comment by gagan0123** [comment]
@MiteshShah 

Also, the current URL is still pointing to the old commit file
`https://rt.cx/psat`

**Comment by gagan0123** [comment]
> @MiteshShah
> 
> Also, the current URL is still pointing to the old commit file `https://rt.cx/psat`

It's working, must be caching issue in browser, so skip this comment.

File: bin/chrome_launcher.sh (hunk 5 above)
Comment (gagan0123) [reply bin/chrome_launcher.sh:6]: @MiteshShah 
Can this be handled?

**Comment by gagan0123** [comment]
@MiteshShah 

I added two commits for:
1. [Go back to the initial directory you started the command in](https://github.com/GoogleChromeLabs/ps-analysis-tool/pull/211/commits/b7a6f9a7b27cf5bdb47ef8c24b8e805cf9bd0877)
1. [Start Chrome maximized and open DevTools automatically](https://github.com/GoogleChromeLabs/ps-analysis-tool/pull/211/commits/305955b42c7fbc80a995445be593245832cfaa8c)

Can you please review the code, test it out and improve if needed?

**Comment by MiteshShah** [comment]
@gagan0123 All looks good to me.

File: bin/chrome_launcher.sh [hunk 6]
```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+
+# Download Extension
+extension_setup() {
+  ps_analysis_tool_version=v0.3.1
```
Comment (mohdsayed) [inline bin/chrome_launcher.sh:5]: v0.3.2 just released today

Comment (MiteshShah) [reply bin/chrome_launcher.sh:5]: Done!

**Comment by gagan0123** [comment]
@MiteshShah 

When using MacOS, the directory `/private/tmp` is utilized to download the extension. However, this directory has a feature that deletes files that have not been used in a while, which results in the deletion of extension files. Even though the directory structure remains in place, all files within it are deleted. Currently, we only verify the existence of the directory, which returns true even when all the extension files have been deleted. Therefore, we need to relocate the unzipped extension to a different temporary location for MacOS. I think `/var/tmp` will work for our usecase. Do you have any suggestions?

**Comment by MiteshShah** [comment]
@gagan0123 we can use `/var/tmp` or we can use `~/.psat` inside user home directory.
Let me know what you prefer and I'll update the script.

**Comment by MiteshShah** [comment]
@gagan0123 we can use `/var/tmp` or we can use `~/.psat` inside user home directory.
Let me know what you prefer and I'll update the script.

**Review by gagan0123 — APPROVED** [review]

**Comment by gagan0123** [comment]
@mohdsayed 

After the script is installed, we now show a message that you have to either re-source the `~/.bashrc` or `~/.zshrc` based on the shell the user has, and provide the list of commands as well as shown in the following screenshot:
image

Do you have any suggestions to improve the output or to make it even clearer to the user, in terms of instructions, what the next steps are after installing the script?

**Comment by mohdsayed** [comment]
> @mohdsayed
> 
> After the script is installed, we now show a message that you have to either re-source the `~/.bashrc` or `~/.zshrc` based on the shell the user has, and provide the list of commands as well as shown in the following screenshot: image
> 
> Do you have any suggestions to improve the output or to make it even clearer to the user, in terms of instructions, what the next steps are after installing the script?

I think we can also add some description to the commands, explaining what they do.

**Comment by gagan0123** [comment]
How about the following?
image

**Comment by gagan0123** [comment]
@mohdsayed 

How about the following?
image

**Comment by prasadnevase** [comment]
@gagan0123 this looks good to me. Lets merge as it will also address #409

**Pull request merged**